File: 016-binary-tree-preorder-traversal/Yiqi_Ou.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        Stack<TreeNode> stack = new Stack<>();
        List<Integer> res = new ArrayList<>();
        TreeNode cur = root;
        while(cur != null || !stack.isEmpty()) {
            if(cur != null) {
                res.add(cur.val);
                if(cur.right != null)
                    stack.push(cur.right);
                cur = cur.left;
            }
            else {
                cur = stack.pop();
            }
        }
        return res;
    }
}

File: 016-binary-tree-preorder-traversal/Yiqi_Ou.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    var stack = [], res = [];
    if(root === null)
        return res;
    var cur = root;
    while(cur !== null || stack.length) {
        if(cur !== null) {
            res.push(cur.val);
            if(cur.right !== null)
                stack.push(cur.right);
            cur = cur.left;
        }
        else {
            cur = stack.pop();
        }
    }
    return res;
};